scripts/evaluate_string_segments.py

```diff
@@ -0,0 +1,327 @@
+
+
+def print_error(*args, **kwargs) -> None:
+    print(*args, file=sys.stderr, **kwargs)
+
+
+def output(*args, **kwargs) -> None:
+    print(*args, file=output_file, **kwargs)
+
+def rm_dir(path: str) -> None:
+    dirpath = Path(path)
+    if dirpath.exists() and dirpath.is_dir():
+        shutil.rmtree(dirpath)
+
+@dataclass(frozen=True)
+class Configuration:
+    output_file: str
+    build_path: str
+    tmp_path: str
+    scale_factor: float
+    encoding_under_test: str
+    time_limit: int
+    debug: bool
+
+
+class Benchmark(ABC):
+    _config: Configuration
+
+    # These encodings are supported by default for string types.
+    _encodings = [
+        'Unencoded',
+        'Dictionary',
+        'RunLength',
+        'FixedStringDictionary',
+        'LZ4'
+    ]
+
+    @staticmethod
+    def create(name: str, *args: Any, **kwargs: Any) -> 'Benchmark':
+        classes = {
+            'hyriseBenchmarkTPCH': TPCHBenchmark,
+            'hyriseBenchmarkTPCDS': TPCDSBenchmark,
+            'hyriseBenchmarkJoinOrder': JoinOrderBenchmark,
+            'hyriseBenchmarkStarSchema': StarSchemaBenchmark
+        }
+        return classes[name](*args, **kwargs)
+    
+    def run(self, config: Configuration) -> None:
+        try:
+            self._config = config
+            output(f'## Benchmark {self.name}:')
+            for threading in ['ST', 'MT']:
+                output(f'### {"Single Thread" if threading == "ST" else "Mulitple Threads"}')
+                result_jsons = [self._run(threading, encoding) for encoding in self._encodings]
+                test_json = self._run(threading, self._config.encoding_under_test)
+                for (encoding, json) in zip(self._encodings, result_jsons):
+                    check_command = ['./scripts/compare_benchmarks.py', json, test_json]
+                    compare_result = check_output(check_command)
+                    output(f'Result for {encoding} vs. {self._config.encoding_under_test}:')
+                    output(compare_result.decode(encoding='utf-8'))
+        except Exception as ex:
+            error_message = f'Skipped {self.name} due to error {ex}.'
+            print_error(error_message)
+            output(error_message)
+
+    @abstractmethod
+    def _run(self, threading: Literal['ST', 'MT'], encoding: str) -> str:
+        pass
+
+    @property
+    def _path(self) -> str:
+        pass
+
+    @abstractmethod
+    def _output_path(self, threading: Literal['ST'] | Literal['MT']) -> str:
+        pass
+
+
+class TPCHBenchmark(Benchmark):
+    name = 'tpch'
+    def _run(self, threading: Literal['ST', 'MT'], encoding: str) -> str:
+        rm_dir("tpch_cached_tables")
+        if threading == 'ST':
+            benchmark_path = self._path
+            st_output_path = self._output_path(threading, encoding)
+            st_command = [benchmark_path, '-o', st_output_path, '-t', str(self._config.time_limit), '-s', str(self._config.scale_factor), '-e', encoding]
```

Comment:
Wouldn't it be enough to only use the if else here. The lines above and under it looks the same for me.

Reply:
Sounds good. Initially, I was unsure whether they would need different handling, so I split them up. Will change it.

Reply:
I moved the complete `_run` method to the base class and only left the argument creation as a task for the subclasses. Thanks for the suggestion!